compiler/rustc_lint/src/panic_fmt.rs

```diff
@@ -0,0 +1,112 @@
+    /// which would break this.
+    PANIC_FMT,
+    Warn,
+    "detect braces in single-argument panic!() invocations",
+    report_in_external_macro
+}
+
+declare_lint_pass!(PanicFmt => [PANIC_FMT]);
+
+impl<'tcx> LateLintPass<'tcx> for PanicFmt {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'tcx>) {
+        if let hir::ExprKind::Call(f, [arg]) = &expr.kind {
+            if let &ty::FnDef(def_id, _) = cx.typeck_results().expr_ty(f).kind() {
+                if Some(def_id) == cx.tcx.lang_items().begin_panic_fn()
+                    || Some(def_id) == cx.tcx.lang_items().panic_fn()
+                {
+                    check_panic(cx, f, arg);
+                }
+            }
+        }
+    }
+}
+
+fn check_panic<'tcx>(cx: &LateContext<'tcx>, f: &'tcx hir::Expr<'tcx>, arg: &'tcx hir::Expr<'tcx>) {
+    if let hir::ExprKind::Lit(lit) = &arg.kind {
+        if let ast::LitKind::Str(sym, _) = lit.node {
+            let mut expn = f.span.ctxt().outer_expn_data();
+            if let Some(id) = expn.macro_def_id {
+                if cx.tcx.is_diagnostic_item(sym::std_panic_macro, id)
+                    || cx.tcx.is_diagnostic_item(sym::core_panic_macro, id)
+                {
+                    let s = sym.as_str();
+                    if !s.contains(&['{', '}'][..]) {
+                        return;
+                    }
+                    let s = s.replace("{{", "").replace("}}", "");
+                    let looks_like_placeholder =
+                        s.find('{').map_or(false, |i| s[i + 1..].contains('}'));
+                    // Unwrap another level of macro expansion if this panic!()
+                    // was expanded from assert!() or debug_assert!().
+                    for &assert in &[sym::assert_macro, sym::debug_assert_macro] {
+                        let parent = expn.call_site.ctxt().outer_expn_data();
+                        if parent
+                            .macro_def_id
+                            .map_or(false, |id| cx.tcx.is_diagnostic_item(assert, id))
+                        {
+                            expn = parent;
+                        }
+                    }
+                    if looks_like_placeholder {
+                        cx.struct_span_lint(PANIC_FMT, arg.span.source_callsite(), |lint| {
+                            let mut l = lint.build("Panic message contains an unused formatting placeholder");
+                            l.note("This message is not used as a format string when given without arguments, but will be in a future Rust version");
+                            if expn.call_site.contains(arg.span) {
+                                l.span_suggestion(
+                                    arg.span.shrink_to_hi(),
+                                    "add the missing argument(s)",
+                                    ", argument".into(),
+                                    Applicability::HasPlaceholders,
```

Comment:
We could actively parse the potential fmt string to get the accurate arg count, but this might be follow up work.

Reply:
Re-using the fmt string parser that `format_args!()` uses looks like a lot of work. Would be great though. Then it can properly suggest adding `hello=..` for `{hello}`, etc.

Reply:
That turned out to be easier than I thought:

![image](https://user-images.githubusercontent.com/783247/96638581-a6029380-1320-11eb-9dd4-7ae7caf6974a.png)

:tada: